Geofence max horizontal and vertical distance

dagar opened the pull request:
I'm still planning to physically test this once I get my hex working well.

LorenzMeier commented:
This looks pretty good! @friekopter: This is similar to your proposed concept, but integrated with the existing geofencing approach. Any feedback or time to test?

LorenzMeier commented:
@DonLakeFlyer This is something else for the Safety Config screen - it would be nice to be able to set those two params.

friekopter commented:
@dagar thanks for implementing. Look good. I will definitely test this. 
Maybe next week, weather is getting better in europe.
What happens if fence is violated?

thomasgubler commented:
how do we handle conflicts between the polygon and the range? Do we need an additional param to select which one to use? (I am aware that in the current implementation the function returns if the range is violated regardless of the polygon)

thomasgubler commented on src/modules/navigator/geofence.cpp:
```diff
@@ -118,6 +125,29 @@ bool Geofence::inside(const struct vehicle_global_position_s &global_position,
+			struct home_position_s home;
+			orb_copy(ORB_ID(home_position), _home_position_sub, &home);
+
+			float dist_xy = -1.0f;
```
why is this inside the `if(updated)` ? the range should be checked on every call to this function I guess and not only if home updates
```

dagar replied:
Ah, that doesn't make sense. The dangers of pushing untested code.

I'll store the home position in geofence once it's set, but should it check for an updated home position? Is it valid to change the home position later?

thomasgubler replied:
Yes you should check for updated home positions and update the internal copy.

thomasgubler commented:
> how do we handle conflicts between the polygon and the range? Do we need an additional param to select which one to use? (I am aware that in the current implementation the function returns if the range is violated regardless of the polygon)

I see now that range checking is only enabled `if (max_horizontal_distance > 0 || max_vertical_distance > 0)`
I still think an additional param to select which geofence method to use is more intuitive. An additional param allows the user to switch between the methods without changing the (safety-critical) range params.

dagar commented:
I don't have a strong opinion about adding an additional parameter for mode switching. Geofence already defaults to on, but doesn't do anything else something is set. 

What about something like GF_FAILSAFE, GF_MAXRANGE, GF_MAXDIST or even additional values for GF_ON?

LorenzMeier commented:
Additional values for GF_ON sound good (easier if its a single parameter and not two which might be confusing), as long as we can identify all current uses of the variable and ensure to handle the change correctly.

LorenzMeier commented:
@dagar Let us know if you want further feedback or assistance.

dagar commented:
How about this for GF_ON?

GF_ON
0 - off
1 - geofence file only
2 - range check only
3 - both (default)

thomasgubler commented:
Because most users have no geofence file this makes sense. The default range values need discussion maybe.

dagar commented:
I was thinking disabled by default
GF_MAX_HOR_DIST -1
GF_MAX_VER_DIST -1

thomasgubler commented:
then let's put GF_ON to 0 - off to make it more obvious

LorenzMeier commented:
@dagar Happy to merge once these things have been completed.

dagar commented:
Implemented the different geofence mode settings and rebased.
I renamed GF_ON to GF_MODE because I thought it made more sense with this new usage.

LorenzMeier commented:
Awesome! @dagar Somehow I missed your last change set, would have merged earlier, sorry!

LorenzMeier commented:
@dagar What is the testing level of this? From what I get the functionality is complete, but I think we have a gap in the RTL implementation. @thomasgubler Could you comment and lend a hand?

dagar commented:
I was going to handle RTL separately, but I can do it here. What do you think about any type of geofence violation resulting in RTL? Or should that also be configurable?

I tested an earlier version of this HIL, but some things have changed. I'll test again once I've added RTL and try to get a real flight in.

LorenzMeier commented:
I'm happy to merge this before RTL is in because you just added another fence - or you add RTL, whatever you prefer. Note that full SITL is coming along nicely. I'll have an update on that in a couple of days.

thomasgubler commented:
I did not check the diff again and it's difficult because of the rebase/single large commit. If there are no big changes since my last comment (apart from parameter changes) this can go in.

dagar commented:
@thomasgubler main difference is GF_MODE

The other minor difference is passing in the home_position orb id from navigator. Subscribing in the geofence constructor doesn't work (different thread?). https://github.com/PX4/Firmware/pull/1886/files#diff-78db4f86a73f1d77027f3e7c8aebb183R252

I've tested this HIL (only max distances). If you think this is good to merge now I'll add RTL in a separate PR.

thomasgubler commented on src/modules/navigator/geofence.cpp:
```diff
@@ -118,13 +132,50 @@ bool Geofence::inside(const struct vehicle_global_position_s &global_position,
+		}
+	}
+
 	bool inside_fence = inside_polygon(lat, lon, altitude);
```
Shouldn't there be a second check for param_geofence_mode somewhere here? eg run inside_polygon depending on that parameter?

dagar replied:
I extended the check at the top of inside_polygon() because inside_polygon() is also called directly from MissionFeasibilityChecker::checkGeofence()

thomasgubler replied:
true, missed that

thomasgubler commented:
> The other minor difference is passing in the home_position orb id from navigator. Subscribing in the geofence constructor doesn't work (different thread?). https://github.com/PX4/Firmware/pull/1886/files#diff-78db4f86a73f1d77027f3e7c8aebb183R252

Actually would it be possible to only pass the home_position as argument and handle the subscription outside (as it's done anyway in the navigator). Currently geofence acts more as a library and not as a module with it's own subscriptions.

dagar commented:
That's actually what I did initially, but I needed to wait for a good home position (the first update).

LorenzMeier commented:
@thomasgubler So you feel this is good to merge? @dagar What is the testing level? Did  you get a chance yet to fly this?

thomasgubler commented:
I still don't like that  the library function `Geofence::inside(double lat, double lon, float altitude)` now depends on uorb. Why not providing the home position as argument and make sure to call the function only if the home position is valid. Right now it's not obvious to the caller that you get a 'inside geofence' if your home position is not valid. The original design was that the geofence class loads all geofence data (polygon/parameters) and is provided with the changing data needed to calculate the decision by the caller.
Also, de facto the geofence uses the navigator's home_position subscription. Why not pass the values instead of the subscription?

dagar commented:
Is the only way to check for a valid home_position to wait for the first orb publication?

LorenzMeier commented:
You can just do an orb_copy() and check if the timestamp is valid.

LorenzMeier commented:
@dagar Any news on this one? Are we maybe limited with our sim capabilities here? Have you seen the Linux port effort? It would allow to run all apps as threads - is this something that you might be interested in?
Source:
https://github.com/mcharleb/Firmware

Initial docs:
https://pixhawk.org/modules/linux/start

Getting this up and running entirely and connected to a Sim might make testing *a lot* easier than it is now.

dagar commented:
@thomasgubler Please take a look at the last commit. Geofence now uses the home_position in navigator.

thomasgubler commented:
ok better but why are you passing home_pos in the constructor and not like the global_position and gps_position as an argument? Seems inconsistent to me.

LorenzMeier commented:
The home position is dynamic and might reset mid-flight (one of the next things we'll implement is sending the home position via MAVLink). So passing it in the constructor is probably limiting too much.

thomasgubler commented:
@LorenzMeier it's passed and stored as a pointer which is however a strange design

LorenzMeier commented:
Ok, I've looked at the code. Given that the geofence checker might end up in a different thread I would prefer a copy of the home position as class member. It does eat a bit more ram, but the pointer to a struct in a potentially different thread might cause sadness down the track, as its a non-obvious threading constraint.

thomasgubler commented:
@LorenzMeier so why not passing to inside(...) directly? That should be in the same thread as the caller.

LorenzMeier commented:
Yes, that's also decent.

dagar commented:
Alright I'm passing the home_position to inside() by value now.

LorenzMeier commented:
Thanks! Now back to the testing state: What is the test plan for this one? I'd love to see a physical flight test with it before merging. @tumbili and / or @kd0aij might be able to pitch in here.

thomasgubler commented:
A HIL test should be enough...

dagar commented:
Verified this is working HIL
[18:25:50.618 - COMP:50] Critical: GF violation: flight termination

I'll add another critical mavlink message that's more descriptive.

dagar commented:
I split the horizontal and vertical checks and added mavlink critical messages.
https://github.com/dagar/Firmware/commit/f69c2a3542d4ed3a04b42b4d3f39c7e080c83788#diff-1e765b01094b58cf0b861540db57b237R149
https://github.com/dagar/Firmware/commit/f69c2a3542d4ed3a04b42b4d3f39c7e080c83788#diff-1e765b01094b58cf0b861540db57b237R155
